### PR TITLE
Clang format for `Common++`

### LINKED ISSUE #1248: introducing clang-format
### Formatting applied to:
- [x] Common++
- [x] Packet++ (https://github.com/seladb/PcapPlusPlus/pull/1510)
- [x] Packet++ Tests (#1456)
- [x] Pcap++ (#1516)
- [x] Pcap++ Tests (#1456)
- [x] PcapTestFramework
- [x] Fuzzers
- [x] Examples (#1464)
- [x] Examples Tests (#1456)

### PR DESCRIPTION
**Pull request by tigercosmos**
apply clang-format for `Common++`

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -1,7 +1,8 @@
 ---
 BasedOnStyle: Microsoft
 UseTab: Always
+BreakBeforeBraces: Allman
 AllowShortFunctionsOnASingleLine: All
 NamespaceIndentation: All
-
+SortIncludes: false
```
Comment (tigercosmos) [inline .clang-format:7]: prevent some possible includes' error due to the ordering.

File: .clang-format [hunk 2]
```diff
@@ -1,7 +1,8 @@
 ---
 BasedOnStyle: Microsoft
 UseTab: Always
+BreakBeforeBraces: Allman
```
Comment (tigercosmos) [inline .clang-format:4]: the style like this:
```
void Foo ()
{
// ...
}
```

File: .github/workflows/build_and_test.yml [hunk 3]
```diff
@@ -30,6 +30,7 @@ jobs:
         run: |
           apk update && apk add cppcheck python3-dev
           python3 -m pip install cmake-format
+          pip install clang-format==18.1.6
```
Comment (tigercosmos) [inline .github/workflows/build_and_test.yml:33]: specify the clang-formant here.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -16,9 +16,11 @@ repos:
   - repo: https://github.com/pocc/pre-commit-hooks
     rev: v1.3.5
     hooks:
+      - id: clang-format
+        args: ["--style=file"]  # Use the .clang-format file for configuration
+        files: ^Common\+\+/.*\.(cpp|h)$
```
Comment (tigercosmos) [inline .pre-commit-config.yaml:21]: apply on `Common++` for now.

File: Common++/header/DeprecationUtils.h [hunk 5]
```diff
@@ -3,35 +3,37 @@
...
+#define DISABLE_WARNING_PUSH DO_PRAGMA(GCC diagnostic push)
+#define DISABLE_WARNING_POP DO_PRAGMA(GCC diagnostic pop)
+#define DISABLE_WARNING(warningName) DO_PRAGMA(GCC diagnostic ignored #warningName)
 
-        #define DISABLE_WARNING_DEPRECATED     DISABLE_WARNING(-Wdeprecated-declarations)
-    #else
-        #pragma message("WARNING: Disabling of warnings is not implemented for this compiler")
-        #define DISABLE_WARNING_PUSH
-        #define DISABLE_WARNING_POP
+// clang-format off
```
Comment (tigercosmos) [inline Common++/header/DeprecationUtils.h:29]: here is the manual fix, since the clang-formant cannot do the formating well here. it tries to separate `-Wdeprecated-declarations` to `-Wdeprecated - declarations`

**Comment by tigercosmos** [comment]
don't squash the commit.

**Comment by seladb** [comment]
@tigercosmos the PR title should be:
"Clang format for `Common++`", not `Commit++` 🙂

**Comment by tigercosmos** [comment]
sorry, fixed the title.

btw, seems that the CI already runs the check for clang-format

**Comment by tigercosmos** [comment]
Sorry, I think it's okay to squash the commits, and we can add the ignore hash in the next PR.

File: Common++/header/DeprecationUtils.h [hunk 6]
```diff
@@ -3,35 +3,37 @@
...
+// clang-format on
+#else
+#pragma message("WARNING: Disabling of warnings is not implemented for this compiler")
+#define DISABLE_WARNING_PUSH
+#define DISABLE_WARNING_POP
 
-        #define DISABLE_WARNING_DEPRECATED
-    #endif
+#define DISABLE_WARNING_DEPRECATED
+#endif
```
Comment (Dimi1010) [inline Common++/header/DeprecationUtils.h:38]: Can we add comments to the `endif`s if clang is going to format away the spaces (or disable that formatting?). It was fine without when they were indented, but it gets really hard to follow nested chains when they are flat.

Comment (tigercosmos) [reply Common++/header/DeprecationUtils.h]: @Dimi1010  sure, apply indents for the macros

File: .clang-format [hunk 7]
```diff
@@ -1,7 +1,10 @@
...
 AllowShortFunctionsOnASingleLine: All
 NamespaceIndentation: All
-
+SortIncludes: false
+IndentPPDirectives: AfterHash
```
Comment (tigercosmos) [inline .clang-format:8]: let the macro has indent

File: .clang-format [hunk 8]
```diff
@@ -1,7 +1,10 @@
...
+IndentWidth: 4
```
Comment (tigercosmos) [inline .clang-format:9]: macro indent width

File: Common++/header/Logger.h [hunk 9]
```diff
@@ -52,64 +55,64 @@ namespace pcpp
...
-		NetworkUtils, ///< NetworkUtils module (Pcap++)
+		PcapLogModuleRemoteDevice,		///< WinPcapRemoteDevice module (Pcap++)
+		PcapLogModuleLiveDevice,		///< PcapLiveDevice module (Pcap++)
+		PcapLogModuleFileDevice,		///< FileDevice module (Pcap++)
+		PcapLogModulePfRingDevice,		///< PfRingDevice module (Pcap++)
+		PcapLogModuleMBufRawPacket,		///< MBufRawPacket module (Pcap++)
+		PcapLogModuleDpdkDevice,		///< DpdkDevice module (Pcap++)
+		PcapLogModuleKniDevice,			///< KniDevice module (Pcap++)
+		PcapLogModuleXdpDevice,			///< XdpDevice module (Pcap++)
+		NetworkUtils,					///< NetworkUtils module (Pcap++)
```
Comment (seladb) [inline Common++/header/Logger.h:115]: Also here - I'm not sure why clang-format decided to use tabs instead of spaces. It doesn't look as it should:

image

File: Common++/header/Logger.h [hunk 10]
```diff
@@ -153,7 +158,8 @@ namespace pcpp
 		 * @param[in] method The method in PcapPlusPlus code the log message is coming from
 		 * @param[in] line The line in PcapPlusPlus code the log message is coming from
 		 */
-		typedef void (*LogPrinter)(LogLevel logLevel, const std::string& logMessage, const std::string& file, const std::string& method, const int line);
+		typedef void (*LogPrinter)(LogLevel logLevel, const std::string &logMessage, const std::string &file,
+								   const std::string &method, const int line);
```
Comment (seladb) [inline Common++/header/Logger.h:162]: In this case clang-format is using a combination of spaces and tabs, it's even more weird...

File: Common++/header/Logger.h [hunk 11]
```diff
@@ -187,7 +193,11 @@ namespace pcpp
 		 * Set all PcapPlusPlus modules to a certain log level
 		 * @param[in] level The log level to set all modules to
 		 */
-		void setAllModulesToLogLevel(LogLevel level) { for (int i=1; i<NumOfLogModules; i++) m_LogModulesArray[i] = level; }
+		void setAllModulesToLogLevel(LogLevel level)
+		{
+			for (int i = 1; i < NumOfLogModules; i++)
+				m_LogModulesArray[i] = level;
```
Comment (seladb) [inline Common++/header/Logger.h:199]: Is there a way to always enforce curly braces even if there's only one statement?

File: Common++/src/Logger.cpp [hunk 12]
```diff
@@ -4,55 +4,50 @@
...
-	delete logStream;
-	if (logLevel == Logger::Error)
+	void Logger::defaultLogPrinter(LogLevel logLevel, const std::string &logMessage, const std::string &file,
+								   const std::string &method, const int line)
 	{
-		m_LastError = logMessage;
+		std::ostringstream sstream;
+		sstream << file << ": " << method << ":" << line;
+		std::cout << std::left << "[" << std::setw(5) << Logger::logLevelAsString(logLevel) << ": " << std::setw(45)
+				  << sstream.str() << "] " << logMessage << std::endl;
```
Comment (seladb) [inline Common++/src/Logger.cpp:33]: I actually like the previous version better, but not sure if there is a way to keep it...

File: Common++/src/SystemUtils.cpp [hunk 13]
```diff
@@ -70,341 +70,300 @@ namespace
...
+	const SystemCore SystemCores::Core31 = {0x80000000, 31};
+
+	const SystemCore SystemCores::IdToSystemCore[MAX_NUM_OF_CORES] = {
+		SystemCores::Core0,	 SystemCores::Core1,  SystemCores::Core2,  SystemCores::Core3,	SystemCores::Core4,
+		SystemCores::Core5,	 SystemCores::Core6,  SystemCores::Core7,  SystemCores::Core8,	SystemCores::Core9,
+		SystemCores::Core10, SystemCores::Core11, SystemCores::Core12, SystemCores::Core13, SystemCores::Core14,
+		SystemCores::Core15, SystemCores::Core16, SystemCores::Core17, SystemCores::Core18, SystemCores::Core19,
+		SystemCores::Core20, SystemCores::Core21, SystemCores::Core22, SystemCores::Core23, SystemCores::Core24,
+		SystemCores::Core25, SystemCores::Core26, SystemCores::Core27, SystemCores::Core28, SystemCores::Core29,
+		SystemCores::Core30, SystemCores::Core31};
```
Comment (seladb) [inline Common++/src/SystemUtils.cpp:113]: is there a way to ask clang-format to keep it the way it was, meaning one value in each line?

File: Common++/header/IpAddressUtils.h [hunk 14]
```diff
@@ -21,83 +21,83 @@ namespace pcpp
 	 * Overload of the equal-to operator
 	 * @return True if the addresses are equal, false otherwise
 	 */
-	bool operator==(const IPv4Address& lhs, const in_addr& rhs);
+	bool operator==(const IPv4Address &lhs, const in_addr &rhs);
```
[the remark this answers is not in the document]
Comment (tigercosmos) [reply Common++/header/IpAddressUtils.h:24]: I just follow the default choice. let me change it.

File: Common++/header/GeneralUtils.h [hunk 15]
```diff
@@ -1,7 +1,7 @@
 #pragma once
 
-#include <string>
 #include <stdint.h>
+#include <string>
```
[the remark this answers is not in the document]
Comment (tigercosmos) [reply Common++/header/GeneralUtils.h:4]: I added that option later, but I think it's better to reorder as long as there is no build error?
I made the reordering to off because it may break the rest of the codebase.

File: Common++/header/IpAddress.h [hunk 16]
```diff
@@ -31,7 +30,7 @@ namespace pcpp
 	 */
 	class IPv4Address
 	{
-	public:
+	  public:
```
[the remark this answers is not in the document]
Comment (tigercosmos) [reply Common++/header/IpAddress.h:33]: let me find if it can be addressed.

File: Common++/header/IpAddress.h [hunk 17]
```diff
@@ -93,20 +92,21 @@ namespace pcpp
...
+		bool operator<(const IPv4Address &rhs) const
 		{
 			uint32_t intVal = toInt();
-			std::reverse(reinterpret_cast<uint8_t*>(&intVal), reinterpret_cast<uint8_t*>(&intVal) + sizeof(intVal));
+			std::reverse(reinterpret_cast<uint8_t *>(&intVal), reinterpret_cast<uint8_t *>(&intVal) + sizeof(intVal));
 
 			uint32_t rhsIntVal = rhs.toInt();
-			std::reverse(reinterpret_cast<uint8_t*>(&rhsIntVal), reinterpret_cast<uint8_t*>(&rhsIntVal) + sizeof(rhsIntVal));
+			std::reverse(reinterpret_cast<uint8_t *>(&rhsIntVal),
+						 reinterpret_cast<uint8_t *>(&rhsIntVal) + sizeof(rhsIntVal));
```
[the remark this answers is not in the document]
Comment (tigercosmos) [reply Common++/header/IpAddress.h:109]: unless we change to spaces, otherwise I don' t think we can address this.

File: Common++/src/SystemUtils.cpp (hunk 13 above)
Comment (tigercosmos) [reply Common++/src/SystemUtils.cpp:113]: I agree, this doesn't look good.

File: Common++/header/GeneralUtils.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (seladb) [reply Common++/header/GeneralUtils.h:4]: Once we agree on all the formatting rules, I suggest we re-run the existing Common++ codebase with `clang-format` to make sure the changes match the formatting rules

File: Common++/src/SystemUtils.cpp (hunk 13 above)
Comment (tigercosmos) [reply Common++/src/SystemUtils.cpp:113]: there is `BreakArrays`, but only for JSON https://clang.llvm.org/docs/ClangFormatStyleOptions.html#breakarrays

File: Common++/header/IpAddress.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (seladb) [reply Common++/header/IpAddress.h:109]: Maybe try to use the other options of `UseTab`? I'm not sure what each one does exactly, but maybe one of them addresses this issue:
https://clang.llvm.org/docs/ClangFormatStyleOptions.html#usetab

File: Common++/src/SystemUtils.cpp (hunk 13 above)
Comment (seladb) [reply Common++/src/SystemUtils.cpp]: Maybe `Cpp11BracedListStyle: false`? 🤔 
https://www.reddit.com/r/C_Programming/comments/izosbi/how_to_clangformat_struct_initializer_list/?rdt=38280

File: Common++/header/IpAddressUtils.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (tigercosmos) [reply Common++/header/IpAddressUtils.h]: fixed.

File: Common++/header/GeneralUtils.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (tigercosmos) [reply Common++/header/GeneralUtils.h]: I reset it.

File: Common++/header/IpAddress.h [hunk 18]
```diff
@@ -1,16 +1,15 @@
 #pragma once
 
-#include <stdint.h>
-#include <string.h>
-#include <string>
 #include <algorithm>
-#include <ostream>
 #include <array>
 #include <memory>
+#include <ostream>
```
[the remark this answers is not in the document]
Comment (tigercosmos) [reply Common++/header/IpAddress.h]: fixed.

File: Common++/header/IpAddress.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (tigercosmos) [reply Common++/header/IpAddress.h]: fixed.

File: Common++/header/Logger.h (hunk 9 above)
Comment (tigercosmos) [reply Common++/header/Logger.h]: fixed.

File: Common++/src/Logger.cpp (hunk 12 above)
Comment (tigercosmos) [reply Common++/src/Logger.cpp]: I cannot find the solution for this.

File: Common++/src/SystemUtils.cpp (hunk 13 above)
Comment (tigercosmos) [reply Common++/src/SystemUtils.cpp]: it helps a little bit.
```
	const SystemCore SystemCores::IdToSystemCore[MAX_NUM_OF_CORES] = {
		SystemCores::Core0,  SystemCores::Core1,  SystemCores::Core2,  SystemCores::Core3,  SystemCores::Core4,
		SystemCores::Core5,  SystemCores::Core6,  SystemCores::Core7,  SystemCores::Core8,  SystemCores::Core9,
		SystemCores::Core10, SystemCores::Core11, SystemCores::Core12, SystemCores::Core13, SystemCores::Core14,
		SystemCores::Core15, SystemCores::Core16, SystemCores::Core17, SystemCores::Core18, SystemCores::Core19,
		SystemCores::Core20, SystemCores::Core21, SystemCores::Core22, SystemCores::Core23, SystemCores::Core24,
		SystemCores::Core25, SystemCores::Core26, SystemCores::Core27, SystemCores::Core28, SystemCores::Core29,
		SystemCores::Core30, SystemCores::Core31
	};
```

File: Common++/header/IpAddress.h [hunk 19]
```diff
@@ -41,19 +40,27 @@ namespace pcpp
...
+		}
 
 		/**
 		 * A constructor that creates an instance of the class out of a 4-byte standard array.
 		 * @param[in] bytes The address as 4-byte standard array in network byte order
 		 */
-		IPv4Address(const std::array<uint8_t, 4>& bytes) : m_Bytes(bytes) {}
+		IPv4Address(const std::array<uint8_t, 4>& bytes) : m_Bytes(bytes)
+		{
+		}
```
Comment (seladb) [inline Common++/header/IpAddress.h:63]: This formatting change doesn't make a lot of sense, is there an option to avoid it?

File: Common++/header/LRUList.h [hunk 20]
```diff
@@ -1,10 +1,10 @@
 #pragma once
 
-#include <unordered_map>
 #include <list>
+#include <unordered_map>
```
Comment (seladb) [inline Common++/header/LRUList.h:4]: For some reason it still changes the include order, any idea why?

File: Common++/header/MacAddress.h [hunk 21]
```diff
@@ -50,9 +53,8 @@ namespace pcpp
 		 * If the string doesn't represent a valid MAC address, the constructor throws an exception.
 		 * @param[in] addr the string representing the MAC address in format "00:00:00:00:00:00"
 		 */
-		template<typename T, typename = typename std::enable_if<std::is_convertible<T, std::string>::value>::type>
-		MacAddress(const T& addr) : MacAddress(static_cast<std::string>(addr)) {};
-
+		template <typename T, typename = typename std::enable_if<std::is_convertible<T, std::string>::value>::type>
+		MacAddress(const T& addr) : MacAddress(static_cast<std::string>(addr)){};
```
Comment (seladb) [inline Common++/header/MacAddress.h:57]: Here it keeps the empty curly braces `{}` in the same line, as opposed to `IpAddress.h` where it didn't. Any idea why?

File: Common++/src/OUILookup.cpp [hunk 22]
```diff
@@ -8,103 +8,104 @@
...
+								++ctrRead;
+							}
 						}
 					}
 				}
 			}
+
+			vendorMap.insert({
+			    std::stoull(line.key()), { val["vendor"], vLocalMaskedFilter }
+            });
```
Comment (seladb) [inline Common++/src/OUILookup.cpp:57]: There is a weird mix of tabs and spaces in these lines, but not sure if it can be fixed 🤔

File: Common++/src/SystemUtils.cpp (hunk 13 above)
Comment (seladb) [reply Common++/src/SystemUtils.cpp]: yeah, still doesn't look very nice...

File: . git-blame-ignore-revs
Comment (seladb) [inline . git-blame-ignore-revs:1]: Should we add this as empty file in this commit, or add it later after we merge?

Comment (tigercosmos) [reply . git-blame-ignore-revs:1]: I think we can have it now, we will add the file anyway.

File: Common++/src/OUILookup.cpp (hunk 22 above)
Comment (tigercosmos) [reply Common++/src/OUILookup.cpp:57]: it is in purpose. the indent is tab, the alignment is space.
for example:
```
<tab> void  foo(int a,
<tab><spaces>int b) {}
```

File: Common++/src/SystemUtils.cpp (hunk 13 above)
Comment (tigercosmos) [reply Common++/src/SystemUtils.cpp]: I think it's fine though, actually shorter and cleaner.

File: Common++/header/LRUList.h (hunk 20 above)
Comment (tigercosmos) [reply Common++/header/LRUList.h:4]: yeah, weird.

File: Common++/src/OUILookup.cpp (hunk 22 above)
Comment (seladb) [reply Common++/src/OUILookup.cpp:57]: yeah, but why does line 57 has spaces instead of tabs? Maybe we should just move `});` to line 56?

File: . git-blame-ignore-revs
Comment (seladb) [reply . git-blame-ignore-revs:1]: I don't have a strong opinion, but I think it's better to add the file after we have the commit hash. But if you feel strongly about keeping it, I'm ok with it

File: Common++/header/MacAddress.h (hunk 21 above)
Comment (tigercosmos) [reply Common++/header/MacAddress.h:57]: maybe because it's a template?

File: Common++/src/SystemUtils.cpp (hunk 13 above)
Comment (seladb) [reply Common++/src/SystemUtils.cpp]: ok 👍

File: Common++/src/OUILookup.cpp (hunk 22 above)
Comment (tigercosmos) [reply Common++/src/OUILookup.cpp:57]: yeah... you are right.. it's weird here as well.

File: Common++/header/IpAddress.h (hunk 19 above)
Comment (tigercosmos) [reply Common++/header/IpAddress.h:63]: this one is due to `BreakBeforeBraces: Allman`

File: Common++/header/MacAddress.h (hunk 21 above)
Comment (tigercosmos) [reply Common++/header/MacAddress.h:57]: I think I know why, probably due to `;`

Comment (tigercosmos) [reply Common++/header/MacAddress.h:57]: yeah, after I removed `;`, it becomes normal.
But for now I can do manually fixing, but after when we apply a lot of files, it's hard to pick these syntax issues one by one.

File: Common++/header/IpAddress.h (hunk 19 above)
Comment (seladb) [reply Common++/header/IpAddress.h:63]: If we change to `Custom` we can set `SplitEmptyFunction: false` which I think should solve it:
https://clang.llvm.org/docs/ClangFormatStyleOptions.html#bracewrapping

File: Common++/header/MacAddress.h (hunk 21 above)
Comment (seladb) [reply Common++/header/MacAddress.h]: oh, maybe we should just remove the `;`

File: Common++/header/LRUList.h (hunk 20 above)
Comment (tigercosmos) [reply Common++/header/LRUList.h:4]: I don't know why, but I cannot resolve this one. the only way is to insert empty line to ensure the order not to change, but in this case, it's okay to let it change.

File: Common++/header/MacAddress.h (hunk 21 above)
Comment (seladb) [reply Common++/header/MacAddress.h]: > yeah, after I removed `;`, it becomes normal. But for now I can do manually fixing, but after when we apply a lot of files, it's hard to pick these syntax issues one by one.

we should fix things whenever we see them, of course we might miss some...

File: Common++/header/IpAddress.h (hunk 19 above)
Comment (tigercosmos) [reply Common++/header/IpAddress.h:63]: actually, I would stand for splitting the braces in this case. shall we just use the default `Allman` settings to prevent too much customization from using `Custom`?

File: Common++/src/OUILookup.cpp (hunk 22 above)
Comment (seladb) [reply Common++/src/OUILookup.cpp]: did you try moving `});` from line 57 to line 56 before running `clang-format`?

File: Common++/header/LRUList.h (hunk 20 above)
Comment (seladb) [reply Common++/header/LRUList.h:4]: ok 👍

File: Common++/header/IpAddress.h (hunk 19 above)
Comment (seladb) [reply Common++/header/IpAddress.h:63]: I don't know the default values of the different flags under `BraceWrapping`. If all the defaults are what we need, I think it's ok to set:
```
BreakBeforeBraces: Custom
BraceWrapping:
  SplitEmptyFunction: false
```

File: Common++/src/OUILookup.cpp (hunk 22 above)
Comment (tigercosmos) [reply Common++/src/OUILookup.cpp]: I tried, it doesn't help 
no matter this;
```
vendorMap.insert({std::stoull(line.key()), { val["vendor"], vLocalMaskedFilter }});
```
or
```
			vendorMap.insert(
				{std::stoull(line.key()), { val["vendor"], vLocalMaskedFilter }});
```

they will be formatted as:
```
			vendorMap.insert({
				std::stoull(line.key()), { val["vendor"], vLocalMaskedFilter }
            });
```

one way is to use `// clang-format off` here, but I would suggest just let it be this if it doesn't break the build.
according to my experience, when there is too many braces, clang-format becomes stupid.

File: Common++/header/IpAddress.h (hunk 19 above)
Comment (tigercosmos) [reply Common++/header/IpAddress.h:63]: I tried your config, and the effect is like this:

```
void matchNetwork(const std::string& network) {}
```
become this:
```
void matchNetwork(const std::string& network)
{}
```

File: Common++/src/OUILookup.cpp (hunk 22 above)
Comment (seladb) [reply Common++/src/OUILookup.cpp]: yes, we can just leave it the way it is... 😕

File: Common++/header/IpAddress.h (hunk 19 above)
Comment (seladb) [reply Common++/header/IpAddress.h:63]: this is slightly better I think. The rest of the curly braces formatting stays the same?

Comment (tigercosmos) [reply Common++/header/IpAddress.h:63]: if there is something in the braces, yes, it will be the same.
```
void matchNetwork(const std::string& network)
{
   // ...
}
```

Comment (tigercosmos) [reply Common++/header/IpAddress.h]: sure, applied your config.

**Review by seladb — APPROVED** [review]
